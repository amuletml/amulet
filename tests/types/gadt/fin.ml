type nat = Z | S of nat
type snat 'n =
  | SZ : snat Z
  | SS : snat 'k -> snat (S 'k)
type fin 'n =
  | FZ : fin (S 'k)
  | FS : fin 'k -> fin (S 'k)

let rec to_fin (x : snat 'n) : fin 'n =
  match x with
  | SS SZ -> FS FZ
  | SS x -> FS (to_fin x)
  | SZ -> to_fin SZ
