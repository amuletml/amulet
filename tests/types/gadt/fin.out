fin.ml[11:14 ..11:18]: error
  No instance for 'n ~ S (S 'k) arising in the expression
   │ 
11 │   | SS SZ -> FS FZ
   │              ^^^^^
  • Because:
   │ 
11 │   | SS SZ -> FS FZ
   │              ^^^^^
  Couldn't match actual type 'n ~ 'n
    with the type expected by the context, 'n ~ S (S 'k)
